### DISCUSSION TITLE
A question about pytraj.align()

### DISCUSSION THREAD

**Discussion by 1heman** [Amber-MD/pytraj#1590] [Q&A] [answered]
Hi,
thanks to making-it-rain, I have been introduced to AmberTools and pytraj.
Prior to this, I have been a GROMACS user.

Unlike GROMACS, where we have to perform 2 operations on the trajectory namely centering and fitting, there seems to be only one here: pytraj.align.

I am particularly not sure how accurate this function is. I have been simulating protein-ligand systems. In many cases, in a visualizer, I have found the ligand to be all over the place but the RMSD of the ligand and the protein speak otherwise.

This is how I have been employing the function:
pt.align(traj_load, ref=0)
Note that I have not given any mask assuming that this would align the trajectory w.r.t. all atoms of the initial frame.
I would also be very pleased to know how to select only the atoms of the protein and ligand for this alignment process.

Thank you,
Hemant

**Comment by drroe** [comment]
> I am particularly not sure how accurate this function is. I have been simulating protein-ligand systems. In many cases, in a visualizer, I have found the ligand to be all over the place but the RMSD of the ligand and the protein speak otherwise.

We will be better able to help if you could post your exact input (commands etc).

> I would also be very pleased to know how to select only the atoms of the protein and ligand for this alignment process.

See section 3.2 on Atom mask selection syntax in the [CPPTRAJ manual](https://jenkins.jasonswails.com/job/amber-github/job/cpptraj/job/master/lastSuccessfulBuild/artifact/doc/CpptrajManual.pdf).

**Reply by 1heman** [reply]
I am using a colab notebook created by [making-it-rain](https://github.com/pablo-arantes/making-it-rain/blob/main/Protein_ligand.ipynb)

I am not sure how helpful these commands by themselves will be but here they are:
```
#@title **Concatenate and align the trajectory**
Skip = "1" #@param ["1", "2", "5", "10", "20", "50"]
stride_traj = Skip
Output_format = "dcd" #@param ["dcd", "pdb", "trr", "xtc"]

simulation_time_analysis = stride_time_ps*nstride
simulation_ns = float(Stride_Time)*int(Number_of_strides)
number_frames = int(simulation_time_analysis)/int(Write_the_trajectory)
number_frames_analysis = number_frames/int(stride_traj)


traj_end = os.path.join(workDir, str(Jobname) + "_all.dcd")
traj_end2 = os.path.join(workDir, str(Jobname) + "_all." + str(Output_format))
template =  os.path.join(workDir, str(Jobname) + '_%s.dcd')

flist = [template % str(i) for i in range(1, nstride + 1)]
#print(flist)

trajlist = pt.load(flist, pdb, stride=stride_traj)
traj_image = trajlist.iterframe(autoimage=True, rmsfit=0)
traj_write = pt.write_traj(traj_end, traj_image, overwrite=True)
traj_load = pt.load(traj_end, pdb)
traj_align = pt.align(traj_load, ref=0) 
traj_write = pt.write_traj(traj_end, traj_align, overwrite=True, options='dcd')
traj_write = pt.write_traj(traj_end2, traj_align, overwrite=True, options=Output_format)
traj_load = pt.load(traj_end, os.path.join(workDir, "SYS_gaff2.prmtop"))
print(traj_load)


traj_end_check = os.path.exists(traj_end2)

if traj_end_check == True:
  print("Trajectory concatenated successfully! :-)")
else:
  print("ERROR: Check your inputs! ")
```
There, in the **Load, view and check the trajectory** subsection, the above lines are contained.

The full notebook can be found [here](https://github.com/pablo-arantes/making-it-rain/blob/main/Protein_ligand.ipynb)

**Comment by hainm** [comment, marked as answer]
hi @1heman: I am sorry for my delay reply (I normally check github notification via my email but github did not seem to send this thread to my email).

> I am particularly not sure how accurate this function is. I have been simulating protein-ligand systems. In many cases, in a > visualizer, I have found the ligand to be all over the place but the RMSD of the ligand and the protein speak otherwise.

To make sure the code correct, please specify a protein segment, saying residue `:1-12`, write the aligned trajectory and visualize it.

> pt.align(traj_load, ref=0, mask=':1-12')

> Note that I have not given any mask assuming that this would align the trajectory w.r.t. all atoms of the initial frame.

Yes, all atoms.

**Comment by hainm** [comment]
@1heman: FYI`pytraj` does have `center` function (but like you said, `pytraj.align` (cpptraj under the hood) combines both `center` and `fitting`)

https://github.com/Amber-MD/pytraj/blob/30d43ce73937b241b21ec08ef12fb013229d099c/pytraj/all_actions.py#L2164-L2206
